src/mo_frontend/typing.ml

```diff
@@ -280,15 +300,92 @@ let detect_unused env inner_identifiers =
+let enter_scope env =
+  let used_identifiers_before = !(env.used_identifiers) in
+  let generic_count_before = !(env.generic_variable_count) in
+  let captured_before = !(env.captured) in
+  let parameters_before = !(env.parameters) in
+  (used_identifiers_before, generic_count_before, captured_before, parameters_before)
 
 let leave_scope env inner_identifiers initial_usage =
+  let used_identifiers_before, generic_count_before, captured_before, parameters_before = initial_usage in
   detect_unused env inner_identifiers;
   let inner_identifiers = get_identifiers inner_identifiers in
   let unshadowed_usage = S.diff !(env.used_identifiers) inner_identifiers in
-  let final_usage = S.union initial_usage unshadowed_usage in
-  env.used_identifiers := final_usage
+  let final_usage = S.union used_identifiers_before unshadowed_usage in
+  env.used_identifiers := final_usage;
+  env.generic_variable_count := generic_count_before;
+  let unshadowed_captured = S.diff !(env.captured) inner_identifiers in
+  env.captured := S.union captured_before unshadowed_captured;
+  env.parameters := parameters_before
+
+(* Stable functions support *)
+
+let captured_variable_name env id =
+  let parameters = !(env.parameters) in
+  match T.Env.find_opt id parameters.positions with
+  | Some position -> Printf.sprintf "<param-%i>" position;
+  | None -> id
+
+let collect_captured_variables env =
+  let captured = List.filter_map (fun id ->
+    match T.Env.find_opt id env.vals with
+    | Some (typ, _, _, _, Nested) -> 
+      let stable_name = captured_variable_name env id in
+      let captured_variable = Type.{ stable_name; variable_type = typ } in
+      Some (id, captured_variable)
+    | _ -> None
+  ) (S.elements !(env.captured)) in
+  T.Env.from_list captured
+
+let stable_function_closure env named_scope =
+  match named_scope with
+  | None -> None
+  | Some function_path ->
+    let captured_variables = collect_captured_variables env in
+    Some T.{
+      function_path;
+      captured_variables;
+    }
+
+let enter_named_scope env name =
+  if (String.contains name '@') || (String.contains name '$') || not !Mo_config.Flags.enhanced_orthogonal_persistence then
+    None
+  else 
+    (match env.named_scope with
+    | None -> None
+    | Some prefix -> Some (prefix @ [name]))
+
+(* Within each scope, generic type parameters get a unique index by 
+   order of definition. This index serves for checking the compatibility 
+   of stable closures that capture generic type parameters.
+   For nested generic classes and functions, the numbering is continued
+   in the sub-scopes. *)
+let generic_variable_index env =
+  let index = !(env.generic_variable_count) in
+  env.generic_variable_count := index + 1;
+  Some index
+
+let append_parameter parameters pattern =
+  match pattern.it with
+  | VarP id | AnnotP ({it = VarP id; _}, _) ->
+    let { count; positions } = parameters in
+    let positions = T.Env.remove id.it positions in
```

Comment:
```suggestion
    let positions = T.Env.remove id.it positions in
```
I don't think there's any need to remove, is there?

Reply:
Yes, redundant, as `add` is insert or replace :)